lib/matplotlib/backend_bases.py

```diff
@@ -3195,3 +3207,560 @@ def zoom(self, *args):
+        """
+
+        for name, tool in six.iteritems(tools):
+            self.add_tool(name, tool)
+
+    def add_tool(self, name, tool, *args, **kwargs):
+        """
+        Add tool to `NavigationBase`
+
+        Add a tool to the tools controlled by Navigation
+
+        If successful adds a new event `tool_trigger_name` where **name** is
+        the **name** of the tool, this event is fired everytime
+        the tool is triggered.
+
+        Parameters
+        ----------
+        name : str
+            Name of the tool, treated as the ID, has to be unique
+        tool : class_like, i.e. str or type
+            Reference to find the class of the Tool to added.
+
+        Notes
+        -----
+        args and kwargs get passed directly to the tools constructor.
+
+        See Also
+        --------
+        matplotlib.backend_tools.ToolBase : The base class for tools.
+        """
+
+        tool_cls = self._get_cls_to_instantiate(tool)
+        if tool_cls is False:
+            raise ValueError('Impossible to find class for %s' % str(tool))
+
+        if name in self._tools:
+            warnings.warn('A "Tool class" with the same name already exists, '
+                          'not added')
+            return self._tools[name]
+
+        self._tools[name] = tool_cls(self, name, *args, **kwargs)
+
+        if tool_cls.keymap is not None:
+            self.set_tool_keymap(name, tool_cls.keymap)
+
+        # For toggle tools init the radio_group in self._toggled
+        if isinstance(self._tools[name], tools.ToolToggleBase):
+            # None group is not mutually exclusive, a set is used to keep track
+            # of all toggled tools in this group
+            if tool_cls.radio_group is None:
+                self._toggled.setdefault(None, set())
+            else:
+                self._toggled.setdefault(tool_cls.radio_group, None)
+
+        self._tool_added_event(self._tools[name])
+        return self._tools[name]
+
+    def _tool_added_event(self, tool):
+        s = 'tool_added_event'
+        event = ToolEvent(s, self, tool)
+        self._callbacks.process(s, event)
+
+    def _handle_toggle(self, tool, sender, canvasevent, data):
+        """
+        Toggle tools, need to untoggle prior to using other Toggle tool
+        Called from tool_trigger_event
+        """
+
+        radio_group = tool.radio_group
+        # radio_group None is not mutually exclusive
+        # just keep track of toggled tools in this group
+        if radio_group is None:
+            if tool.toggled:
+                self._toggled[None].remove(tool.name)
+            else:
+                self._toggled[None].add(tool.name)
+            return
+
+        # If the tool already has a toggled state, untoggle it
+        if self._toggled[radio_group] == tool.name:
+            toggled = None
+        # If no tool was toggled in the radio_group
+        # toggle it
+        elif self._toggled.get(radio_group, None) is None:
```

Comment:
Given the code above, `radio_group` ever not be in self.`_toggled`?  Also, if it were to not be, it would have raised in the `if` condition above.